File: src/test/scala/notjoe/stockpile/inventory/MassItemInventorySpec.scala
package notjoe.stockpile.inventory

import net.minecraft.item.{ItemStack, Items}
import net.minecraft.util.registry.Registry
import notjoe.stockpile.test.TestItems
import org.junit.runner.RunWith
import org.powermock.core.classloader.annotations.{PrepareForTest, SuppressStaticInitializationFor}
import org.powermock.modules.junit4.{PowerMockRunner, PowerMockRunnerDelegate}
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FlatSpec, Matchers}

@RunWith(classOf[PowerMockRunner])
@PowerMockRunnerDelegate(classOf[JUnitRunner])
@SuppressStaticInitializationFor(Array("net.minecraft.item.Items", "net.minecraft.util.registry.Registry"))
@PrepareForTest(Array(classOf[Items], classOf[Registry[_]]))
class MassItemInventorySpec extends FlatSpec with Matchers {
  "A MassItemInventory" should "accept any item when first created" in {
    val inventory = new MassItemInventory()
    inventory.isAcceptingNewStackType shouldBe true
  }

  it should "be empty when first created" in {
    val inventory = new MassItemInventory()
    inventory.isInvEmpty shouldBe true
  }

  it should "update its stored stack type when the first item is inserted" in {
    val inventory = new MassItemInventory()
    val insertedStack = new ItemStack(TestItems.Red, 1)

    inventory.insertStack(insertedStack)

    ItemStack.areEqual(inventory.stackType, insertedStack) shouldBe true
  }

  it should "not be empty after the first item is inserted" in {
    val inventory = new MassItemInventory()
    val insertedStack = new ItemStack(TestItems.Red, 1)

    inventory.insertStack(insertedStack)

    inventory.isInvEmpty shouldBe false
  }

  it should "not accept an item of a different type than what's stored inside" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), amountStored = 1)

    val remainder = inventory.insertStack(new ItemStack(TestItems.Blue, 1))

    ItemStack.areEqual(remainder, new ItemStack(TestItems.Blue, 1)) shouldBe true
  }

  it should "not accept items past its capacity" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), maxStacks = 2, amountStored = 64 + 63)

    val remainder = inventory.insertStack(new ItemStack(TestItems.Red, 2))

    ItemStack.areEqual(remainder, new ItemStack(TestItems.Red, 1)) shouldBe true
  }

  it should "allow a new item type when emptied and set to accept new stacks" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), amountStored = 1, allowNewStackWhenEmpty = true)

    inventory.removeInvStack(MassItemInventory.OutputSlotIndex)

    val remainder = inventory.insertStack(new ItemStack(TestItems.Blue, 1))

    ItemStack.areEqual(inventory.stackType, new ItemStack(TestItems.Blue, 1)) shouldBe true
    remainder.isEmpty shouldBe true
  }

  it should "not allow a new item type when emptied and not set to accept new stacks" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), amountStored = 1, allowNewStackWhenEmpty = false)

    inventory.removeInvStack(MassItemInventory.OutputSlotIndex)

    val remainder = inventory.insertStack(new ItemStack(TestItems.Blue, 1))

    ItemStack.areEqual(remainder, new ItemStack(TestItems.Blue, 1)) shouldBe true
    ItemStack.areEqual(inventory.stackType, new ItemStack(TestItems.Red, 1)) shouldBe true
  }

  it should "never report items as valid for its output slot" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), amountStored = 1)
    inventory.isValidInvStack(MassItemInventory.OutputSlotIndex, new ItemStack(TestItems.Red)) shouldBe false
    inventory.isValidInvStack(MassItemInventory.OutputSlotIndex, new ItemStack(TestItems.Blue)) shouldBe false
  }

  it should "report only stacks with the same item type as valid" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), amountStored = 1)
    inventory.isValidInvStack(MassItemInventory.InputSlotIndex, new ItemStack(TestItems.Red)) shouldBe true
    inventory.isValidInvStack(MassItemInventory.InputSlotIndex, new ItemStack(TestItems.Blue)) shouldBe false
  }

  it should "only start to fill its input slot once it's within the last stack of its capacity" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), amountStored = 64, maxStacks = 2)
    inventory.getInvStack(MassItemInventory.InputSlotIndex).isEmpty shouldBe true

    inventory.insertStack(new ItemStack(TestItems.Red, 3))

    ItemStack.areEqual(inventory.getInvStack(MassItemInventory.InputSlotIndex), new ItemStack(TestItems.Red, 3)) shouldBe true
  }

  it should "lower its stored amount when stacks are taken" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), amountStored = 128)

    inventory.removeInvStack(MassItemInventory.OutputSlotIndex)

    inventory.amountStored shouldBe 64
  }

  it should "set its stored amount to 0 and stack type to empty when cleared" in {
    val inventory = new MassItemInventory(_stackType =  new ItemStack(TestItems.Red), amountStored = 1)
    inventory.method_5448()

    inventory.amountStored shouldBe 0
    inventory.stackType.isEmpty shouldBe true
  }
}

File: src/test/kotlin/me/branchpanic/mods/stockpile/api/inventory/MassItemInventorySpec.kt
package me.branchpanic.mods.stockpile.api.inventory

import io.kotlintest.runner.junit4.KotlinTestRunner
import io.kotlintest.shouldBe
import io.kotlintest.specs.WordSpec
import me.branchpanic.mods.stockpile.api.TestItems
import me.branchpanic.mods.stockpile.api.storage.MassItemStorage
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.util.registry.Registry
import org.junit.runner.RunWith
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.core.classloader.annotations.SuppressStaticInitializationFor
import org.powermock.modules.junit4.PowerMockRunner
import org.powermock.modules.junit4.PowerMockRunnerDelegate

@RunWith(PowerMockRunner::class)
@PowerMockRunnerDelegate(KotlinTestRunner::class)
@SuppressStaticInitializationFor("net.minecraft.item.Items", "net.minecraft.util.registry.Registry")
@PrepareForTest(Items::class, Registry::class)
class MassItemInventorySpec : WordSpec({
    "A mass ItemStack Inventory" should {
        "only accept the item contained in the associated MassStorage" {
            val inv = MassItemInventory(
                MassItemStorage(
                    maxStacks = 1,
                    storedItems = 0L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = true
                )
            )

            inv.isValidInvStack(MassItemInventory.INPUT_SLOT, ItemStack(TestItems.StandardItemA)) shouldBe true
            inv.canInsertInvStack(MassItemInventory.INPUT_SLOT, ItemStack(TestItems.StandardItemA), null) shouldBe true

            inv.isValidInvStack(MassItemInventory.INPUT_SLOT, ItemStack(TestItems.StandardItemB)) shouldBe false
            inv.canInsertInvStack(MassItemInventory.INPUT_SLOT, ItemStack(TestItems.StandardItemB), null) shouldBe false
        }

        "only accept items in its designated input slot" {
            val inv = MassItemInventory(
                MassItemStorage(
                    maxStacks = 1,
                    storedItems = 0L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = true
                )
            )

            inv.isValidInvStack(MassItemInventory.INPUT_SLOT, ItemStack(TestItems.StandardItemA)) shouldBe true
            inv.canInsertInvStack(MassItemInventory.INPUT_SLOT, ItemStack(TestItems.StandardItemA), null) shouldBe true

            inv.isValidInvStack(MassItemInventory.OUTPUT_SLOT, ItemStack(TestItems.StandardItemA)) shouldBe false
            inv.canInsertInvStack(MassItemInventory.OUTPUT_SLOT, ItemStack(TestItems.StandardItemA), null) shouldBe false
        }

        "update the MassStorage's stored item through insertion" {
            val inv = MassItemInventory(
                MassItemStorage(
                    maxStacks = 1,
                    storedItems = 0L,
                    storedStack = ItemStack.EMPTY,
                    clearWhenEmpty = true
                )
            )

            inv.storage.instanceIsSet shouldBe false

            inv.setInvStack(MassItemInventory.INPUT_SLOT, ItemStack(TestItems.StandardItemA, 1))

            inv.storage.instanceIsSet shouldBe true
            inv.storage.currentInstance.item shouldBe TestItems.StandardItemA
            inv.storage.amountStored shouldBe 1L
        }

        "update the MassStorage's stored item through extraction" {
            val inv = MassItemInventory(
                MassItemStorage(
                    maxStacks = 1,
                    storedItems = 64L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = true
                )
            )

            inv.storage.instanceIsSet shouldBe true

            inv.takeInvStack(MassItemInventory.OUTPUT_SLOT, 64)


            inv.storage.instanceIsSet shouldBe false
            inv.storage.currentInstance.isEmpty shouldBe true
            inv.storage.amountStored shouldBe 0L
        }

        "yield one stack's worth of contents in the output slot" {
            MassItemInventory(
                MassItemStorage(
                    maxStacks = 1,
                    storedItems = 32L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = true
                )
            ).getInvStack(MassItemInventory.OUTPUT_SLOT).amount shouldBe 32

            MassItemInventory(
                MassItemStorage(
                    maxStacks = 2,
                    storedItems = 128L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = true
                )
            ).getInvStack(MassItemInventory.OUTPUT_SLOT).amount shouldBe 64

            MassItemInventory(
                MassItemStorage(
                    maxStacks = 1,
                    storedItems = 0L,
                    storedStack = ItemStack.EMPTY,
                    clearWhenEmpty = true
                )
            ).getInvStack(MassItemInventory.OUTPUT_SLOT).amount shouldBe 0
        }

        "start filling the input slot only when within one stack of max capacity" {
            MassItemInventory(
                MassItemStorage(
                    maxStacks = 1,
                    storedItems = 32L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = true
                )
            ).getInvStack(MassItemInventory.INPUT_SLOT).amount shouldBe 0

            MassItemInventory(
                MassItemStorage(
                    maxStacks = 2,
                    storedItems = 128L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = true
                )
            ).getInvStack(MassItemInventory.INPUT_SLOT).amount shouldBe 64

            MassItemInventory(
                MassItemStorage(
                    maxStacks = 2,
                    storedItems = 120L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = true
                )
            ).getInvStack(MassItemInventory.INPUT_SLOT).amount shouldBe 56
        }

        "remove items from the MassStorage through takeInvStack" {
            val inv = MassItemInventory(
                MassItemStorage(
                    maxStacks = 2,
                    storedItems = 128L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = false
                )
            )

            inv.takeInvStack(MassItemInventory.OUTPUT_SLOT, 64).apply {
                item shouldBe TestItems.StandardItemA
                amount shouldBe 64
            }

            inv.storage.amountStored shouldBe 64L

            inv.takeInvStack(MassItemInventory.OUTPUT_SLOT, 32).apply {
                item shouldBe TestItems.StandardItemA
                amount shouldBe 32
            }

            inv.storage.amountStored shouldBe 32L

            inv.takeInvStack(MassItemInventory.OUTPUT_SLOT, 64).apply {
                item shouldBe TestItems.StandardItemA
                amount shouldBe 32
            }

            inv.storage.amountStored shouldBe 0L
        }

        "remove items from the MassStorage through remove" {
            val inv = MassItemInventory(
                MassItemStorage(
                    maxStacks = 2,
                    storedItems = 128L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = false
                )
            )

            inv.removeInvStack(MassItemInventory.OUTPUT_SLOT).apply {
                item shouldBe TestItems.StandardItemA
                amount shouldBe 64
            }

            inv.storage.amountStored shouldBe 64L

            inv.removeInvStack(MassItemInventory.OUTPUT_SLOT).apply {
                item shouldBe TestItems.StandardItemA
                amount shouldBe 64
            }

            inv.storage.amountStored shouldBe 0L
        }

        "add items to the MassStorage dangerously via setInvStack" {
            val inv = MassItemInventory(
                MassItemStorage(
                    maxStacks = 1,
                    storedItems = 0L,
                    storedStack = ItemStack(TestItems.StandardItemA),
                    clearWhenEmpty = false
                )
            )

            inv.setInvStack(MassItemInventory.INPUT_SLOT, ItemStack(TestItems.StandardItemA, 64))

            inv.storage.amountStored shouldBe 64L
        }
    }
})